File: app/src/main/java/com/bt/barrett/sbsr/ReaderActivity.java
package com.bt.barrett.sbsr;

import android.app.Activity;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;

import java.util.LinkedList;
import java.util.ListIterator;

public class ReaderActivity extends Activity {

    TextView sentence;
    Button left, right;
    LinkedList<String> words;
    ListIterator<String> wordCursor;
    String paragraph;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        this.setContentView(R.layout.reader);

        Bundle b = getIntent().getExtras();
        paragraph = b.getString("paragraph");

        initialize();
    }

    private void initialize() {
        sentence = findViewById(R.id.sentence_id);
        left = findViewById(R.id.leftButtonID);
        right = findViewById(R.id.rightButtonID);

        words = new LinkedList<>();

        paragraph = paragraph.replaceAll("[\n]+", ".");
        paragraph = paragraph.replaceAll("\\?", "**Q?");
        paragraph = paragraph.replaceAll("!", "**E!");
        paragraph = paragraph.replaceAll("\\.", "**P.");

        String wordArray[] = paragraph.split("[.?!]");

        for (int i = 0; i < wordArray.length; i++) {

            words.addLast(replace(wordArray[i]));
        }

        wordCursor = words.listIterator();

//        System.out.println(wordArray[0]);

        sentence.setText(wordCursor.next());
//        left.setVisibility(View.INVISIBLE);
//        right.setVisibility(View.INVISIBLE);
    }

    public void onClick(View v) {
        if (v == left) {
            if (wordCursor.hasPrevious())
                sentence.setText(wordCursor.previous());
//            System.out.println("left");
        } else if (v == right) {
            if (wordCursor.hasNext())
                sentence.setText(wordCursor.next());
//            System.out.println("right");
        }
    }

    /*
    *
    * */
    public void next(View v) {
        if (wordCursor.hasNext())
            sentence.setText(wordCursor.next());
//        wordCursor.next();
    }

    public void previous(View v) {
        if (wordCursor.hasPrevious())
            sentence.setText(wordCursor.previous());
//        wordCursor.previous();
    }

    private String replace(String s) {

        String fixedString = s.replaceAll("\\*\\*Q", "?");
        fixedString = fixedString.replaceAll("\\*\\*E", "!");
        fixedString = fixedString.replaceAll("\\*\\*P", ".");

        return fixedString;
    }

}

File: app/src/main/java/com/bt/barrett/sbsr/ReaderActivity.kt
package com.bt.barrett.sbsr

import android.app.Activity
import android.os.Bundle
import android.view.View
import android.widget.Button
import android.widget.TextView
import kotlinx.android.synthetic.main.reader.*
import java.util.*

class ReaderActivity : Activity() {

    private var words: LinkedList<String>? = null
    private var wordCursor: ListIterator<String>? = null
    private var paragraph: String? = null

    public override fun onCreate(savedInstanceState: Bundle) {
        super.onCreate(savedInstanceState)
        this.setContentView(R.layout.reader)
        val b = intent.extras
        paragraph = b.getString("paragraph")
        initialize()
    }

    private fun initialize() {
        words = LinkedList()
        paragraph = paragraph!!.replace("[\n]+".toRegex(), ".")
        paragraph = paragraph!!.replace("\\?".toRegex(), "**Q?")
        paragraph = paragraph!!.replace("!".toRegex(), "**E!")
        paragraph = paragraph!!.replace("\\.".toRegex(), "**P.")
        val wordArray = paragraph!!.split("[.?!]").toTypedArray()
        for (i in wordArray.indices) {
            words!!.addLast(replace(wordArray[i]))
        }
        wordCursor = words!!.listIterator()

//        System.out.println(wordArray[0]);
        sentence.text = (wordCursor as MutableListIterator<String>).next()
//        left.setVisibility(View.INVISIBLE);
//        right.setVisibility(View.INVISIBLE);
    }

    fun onClick(v: View) {
        if (v === leftButtonID) {
            if (wordCursor!!.hasPrevious()) {
                sentence!!.text = wordCursor!!.previous()
            }
        } else if (v === rightButtonID) {
            if (wordCursor!!.hasNext()) {
                sentence!!.text = wordCursor!!.next()
            }
        }
    }

    /*
    *
    * */
    fun next(v: View?) {
        if (wordCursor!!.hasNext()) sentence!!.text = wordCursor!!.next()
        //        wordCursor.next();
    }

    fun previous(v: View?) {
        if (wordCursor!!.hasPrevious()) sentence!!.text = wordCursor!!.previous()
        //        wordCursor.previous();
    }

    private fun replace(s: String): String {
        return s.replace("\\*\\*Q".toRegex(), "?")
                .replace("\\*\\*E".toRegex(), "!")
                .replace("\\*\\*P".toRegex(), ".")
    }
}